### PR TITLE
Implement OpenTracing-like tracer

### LINKED ISSUE #15: Integrate Tracing (derived from OpenTelemetry)
### Specification

OpenTelemetry is an overly complicated beast. It's far too complex to adopt into a logging system. However the basic principles of tracing makes sense. Here I'm showing how you can set one up for comparison testing, for us to derive a tracing schema and later visualise it ourselves or by passing it into an OTLP compatible visualiser.

```sh
docker run -d --name jaeger \
  -e COLLECTOR_ZIPKIN_HOST_PORT=:9411 \
  -e COLLECTOR_OTLP_ENABLED=true \
  -p 6831:6831/udp \
  -p 6832:6832/udp \
  -p 5778:5778 \
  -p 16686:16686 \
  -p 4317:4317 \
  -p 4318:4318 \
  -p 14250:14250 \
  -p 14268:14268 \
  -p 14269:14269 \
  -p 9411:9411 \
  jaegertracing/all-in-one:1.36
```

The above command runs jaeger. Take note of 4318 port which is the OTLP protocol over HTTP.

Visit `localhost:16686` to be able to view the jaeger system.

Then any example code, like for example [https://github.com/open-telemetry/opentelemetry-js/blob/main/examples/basic-tracer-node/index.js](https://github.com/open-telemetry/opentelemetry-js/blob/main/examples/basic-tracer-node/index.js) can run and push traces directly to the docker container.

What is frustrating is:

1. OpenTelemetry code only exports to stderr as an afterthought, it's not considered first class usage
2. The stderr exporters output via `console.log` and produce pretty printed results that are not actual JSON. Thus you cannot just pipe it to a relevant location.
3. The schema of the span data isn't clear, it seems different parts of the documentation still have old data, or maybe the JS implementation itself is hasn't been updated to the new schema.

The plan:

1. Create your own "span" derived from opentelemetry and output as just regular structured JSON
2. Massage it to be compatible to open telemetry viewers like jaeger
3. Use jaeger's 4318 to stream the JSON and view data in the interim
4. Find an easier way to visualise traces, maybe something that can be used CLI or in the GUI
5. For production usage, feed to any structured log capturer, and then feed into a viewer that understands trace information

### Additional context

* #14
* [https://github.com/MatrixAI/js-logger/issues/3#issuecomment-1165576246](https://github.com/MatrixAI/js-logger/issues/3#issuecomment-1165576246)
* [https://github.com/MatrixAI/js-logger/issues/3#issuecomment-1159965066](https://github.com/MatrixAI/js-logger/issues/3#issuecomment-1159965066)
* [https://github.com/MatrixAI/js-logger/issues/3#issuecomment-1158628130](https://github.com/MatrixAI/js-logger/issues/3#issuecomment-1158628130)
* [ENG-100](https://linear.app/matrix-ai/issue/ENG-100)

### Tasks

1. ...
2. ...
3. ...

### PR DESCRIPTION
**Pull request by Abby010**
### Description

This program visualizes real-time execution flow of distributed spans in a terminal-based UI (TUI) using React Ink. It allows users to toggle between time-based (--sample 1s) and logical event-based (--sample logical) views while displaying structured span hierarchies with box-drawing characters.

### Issues Fixed

* [Issue-15](https://github.com/MatrixAI/js-logger/issues/15), Fixes ENG-234
* Fixes #15 

### Tasks

- [x] 1. We now have a working visualization (**Attaching terminal recordings for reference at the end**).
- [x] 2. The core library has been implemented

### Final checklist

* [x] Domain specific tests
* [x] Full tests
* [x] Updated inline-comment documentation
* [x] Lint fixed
* [x] Squash and rebased
* [x] Sanity check the final build

## Terminal 1: --sample logical

- Preview (https://asciinema.org/a/rsIX7rt6teGgXFUyOJsLctsoR)
- Runs cli.tsx, utilizing React-Ink to display spans in a hierarchical tree structure based on logical parent-child relationships.
- Organizes spans according to their causal dependencies (e.g., child spans appear under their respective parent spans).
- Ensures that spans are grouped logically rather than by strict timestamp order, helping to visualize the structured execution flow.

## Terminal 2: --sample 1s

- Preview (https://asciinema.org/a/4u3ZkqPQu1i40ZfzlJ89Pyl91)
- Runs cli.tsx, utilizing React-Ink to display spans sorted by their start time rather than logical hierarchy.
- Orders spans sequentially, meaning spans that start earlier appear first, regardless of parent-child relationships.
- Provides a time-centric view, useful for debugging latency and analyzing performance bottlenecks in real-time execution.

### CODE REVIEW

**Comment by CMCDragonkai** [comment]
Any new visualisations based on lots of lots of data??? @Abby010 and make sure you are copying over and synthesizing the details I have earlier.

**Comment by Abby010** [comment]
> Any new visualisations based on lots of lots of data??? @Abby010 and make sure you are copying over and synthesizing the details I have earlier.

Not yet, I starting getting some dependency errors so I needed to fix that but I have started working with lots of data.

**Comment by Abby010** [comment]
New video links have been added with more data in the description

File: spans.json
Comment (aryanjassal) [inline spans.json:1]: Make sure that you run the following commands before pushing.

- `npm run build`
- `npm run lint`
- `npm run test`

You can omit the first command in most cases if either linting or tests pass. The build command basically reveals any typescript errors which are caused by violating the typescript rules, meaning the program can't be compiled.

If you pushed something in which the CI is failing, then push up the next commit to address the failed CI. This fixing commit should be rebased and squashed into the previous commit, so they look like a single commit and keep the history clean.

Try to ensure the CI is passing at all times. If it is not, and it is intended to not pass, then add the following in a new line in the commit message to skip the CI workflow. You normally do it in WIP commits where it is expected for the CI to fail, and try to avoid skipping CI unless you have a good reason to do so.

```
wip: working on this new feature
[ci skip]
```

The lint command reveals a bunch of linting issues, but problems like incorrect indentation or a missing semicolon can be addressed automatically. Thus, in these cases, prefer to run `npm run lintfix` to resolve automatically fixable lint issues. If there are issues not automatically resolvable, then it will perform all possible fixes and give you a list for manual fixing.

Make sure to resolve all linting errors and warnings. Even a warning will fail the linting step in the CI.

Comment (aryanjassal) [inline spans.json:1]: I'm guessing this file is designed to act as dummy data for testing out your program. Usually, `fast-check` is used for writing tests like this.

Refer to these links for a quick guide on writing tests using fast-check.
- https://github.com/MatrixAI/Polykey/pull/870#discussion_r1970950891
- https://github.com/MatrixAI/Polykey/pull/870#discussion_r1980470956

You can also dynamically generate JSON data using this method. Refer to the implementation here for an example of doing it.
- https://github.com/MatrixAI/js-virtualtar/blob/eac3454ac3b81d781984a06d179bda0fe718f8f8/tests/utils/fastcheck.ts#L126

Refer to this page for a more comprehensive documentation on available arbitraries. We might need to have a meeting to go through using fast-check to generate dummy data.
- https://fast-check.dev/docs/introduction

On the other hand, using pre-set data is also useful for testing. If you're doing a show-case, then using these 'fixtures' makes perfect sense. In that case, it would need to be under a place like `tests/fixtures/span.json`.

Basically, if you're using this for testing, then use fast-check instead. If you're using this for demo, then restructure the file in the project tree.

File: src/Logger.ts [hunk 1]
```diff
@@ -3,6 +3,8 @@ import type Handler from './Handler.js';
 import { LogLevel } from './types.js';
 import ConsoleErrHandler from './handlers/ConsoleErrHandler.js';
 import * as utils from './utils.js';
+import { openSpan, closeSpan } from "./lib/tracingManager.js";
+
```
Comment (aryanjassal) [inline src/Logger.ts]: Usually there is an order to the imports.

```ts
import type { TypeName } from './types';
import DefaultClass from './defaultClass';
import { ClassName } from './class';
import { methodA, methodB } from './methods';
import * as utils from './utils';
```

Types always come first, then default classes, then specific class imports, then specific method imports, then wildcard imports. This can change depending on the source. Importing from other libraries usually has a higher priority than importing from local files. If there are any issues here, the linter will complain, and running a `lintfix` will resolve most of the import order issues.

File: src/Logger.ts [hunk 2]
```diff
@@ -107,89 +109,84 @@ class Logger {
...
-    if (formatOrData == null || typeof formatOrData === 'function') {
-      return this.log(msg, {}, LogLevel.DEBUG, formatOrData as LogFormatter);
-    } else {
-      return this.log(msg, formatOrData, LogLevel.DEBUG, format);
-    }
+    public debug(
+      msg?: ToString,
+      formatOrData?: LogFormatter | LogData,
+      format?: LogFormatter,
+      parentSpanId?: string | undefined
```
Comment (aryanjassal) [inline src/Logger.ts]: Adding a `?` after a parameter name makes it optional. Basically, it can be left undefined. If you have the question mark after the parameter name, you shouldn't need to specify `undefined` as a type.

Also, recheck other parameters to make sure that they are all optional. Basically, make sure the base case of `debug()` will work.

File: src/Logger.ts [hunk 3]
```diff
@@ -107,89 +109,84 @@ class Logger {
...
-  public info(
-    msg: ToString | undefined,
-    data: LogData,
-    format?: LogFormatter,
-  ): void;
+
   public info(
     msg?: ToString,
     formatOrData?: LogFormatter | LogData,
     format?: LogFormatter,
```
Comment (aryanjassal) [inline src/Logger.ts]: Same here. Check if this works on the base case.

File: src/Logger.ts [hunk 4]
```diff
@@ -107,89 +109,84 @@ class Logger {
...
-      return this.log(msg, formatOrData, LogLevel.WARN, format);
-    }
+
+
+    public warn(
+      msg?: ToString,
+      formatOrData?: LogFormatter | LogData,
+      format?: LogFormatter,
+      parentSpanId?: string | undefined
+  ): string {
```
Comment (aryanjassal) [inline src/Logger.ts]: Same here.

File: src/Logger.ts [hunk 5]
```diff
@@ -107,89 +109,84 @@ class Logger {
...
-      return this.log(msg, formatOrData, LogLevel.ERROR, format);
-    }
+
+
+    public error(
+      msg?: ToString,
+      formatOrData?: LogFormatter | LogData,
+      format?: LogFormatter,
+      parentSpanId?: string | undefined
+  ): string {
```
Comment (aryanjassal) [inline src/Logger.ts]: And same here, too.

File: src/bin/SpanTree.tsx [hunk 6]
```diff
@@ -0,0 +1,112 @@
+import React, { FC } from 'react';
+import { Box, Text } from 'ink';
+import { Span } from "../lib/span.js";
```
Comment (aryanjassal) [inline src/bin/SpanTree.tsx]: The linter should have complained about this already. We always use single quotation marks where possible.

File: src/bin/SpanTree.tsx [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+// Props for handling both root and child spans
+interface SpanTreeProps {
+  spans: Span[];
+  sampleMode: string;
+}
+
+/**
+ * Sort spans based on the selected sampling mode.
+ */
+const sortSpans = (spans: Span[], mode: string): Span[] => {
```
Comment (aryanjassal) [inline src/bin/SpanTree.tsx]: If the `mode` parameter has specific available options, then it should be set as such. Also make sure for array types, you use `Array<Type>` instead of `Type[]`.

```ts
interface SpanTreeProps {
  spans: Array<Span>;
  mode: 'logical' | 'time';
};
```

File: src/bin/SpanTree.tsx [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    spans.forEach(span => {
+      if (!spanMap.has(span.spanId)) {
+        const newSpan = new Span(span.name, span.parentSpanId);
+        Object.assign(newSpan, span);
+        newSpan.children = [];
+        spanMap.set(span.spanId, newSpan);
+      }
+    });
+
+    const rootSpans: Span[] = [];
```
Comment (aryanjassal) [inline src/bin/SpanTree.tsx]: Same here.

File: src/bin/SpanTree.tsx [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+ * **Recursive Renderer**
+ * - Uses box-drawing characters (│ ├ └) for structured layout.
+ */
+const RecursiveSpanTree: FC<{ span: Span; prefix: string; isLastChild: boolean }> = ({
+  span,
+  prefix,
+  isLastChild,
+}) => {
+  const connector = isLastChild ? '└── ' : '├── ';
+  const newPrefix = prefix + (isLastChild ? '    ' : '│   '); // Maintain vertical structure
```
Comment (aryanjassal) [inline src/bin/SpanTree.tsx]: Ensure line length (even for comments) doesn't exceed 80 characters.

File: src/bin/TimeLineView.tsx [hunk 10]
```diff
@@ -0,0 +1,111 @@
+// TimelineView.tsx
```
Comment (aryanjassal) [inline src/bin/TimeLineView.tsx]: A file doesn't need to start with the name of the file commented on top.

File: src/bin/TimeLineView.tsx [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+// We make each row = 1000 ms
+const TIME_STEP_MS = 1000;
+
+function timeToRow(timeMs: number, baseTimeMs: number) {
+  return Math.floor((timeMs - baseTimeMs) / TIME_STEP_MS);
+}
+
+function assignLanes(spans: Span[]): Map<string, number> {
```
Comment (aryanjassal) [inline src/bin/TimeLineView.tsx]: Change the type to use Array<>

File: src/bin/TimeLineView.tsx [hunk 12]
```diff
@@ -0,0 +1,111 @@
+// TimelineView.tsx
+import React, { FC } from 'react';
+import { Box, Text } from 'ink';
+
+interface Span {
+  spanId: string;
+  name: string;
+  parentSpanId: string | null;
+  startTime: number;
+  endTime: number | null;
```
Comment (aryanjassal) [inline src/bin/TimeLineView.tsx]: Mostly, instead of using `null`, we prefer to use `undefined` for missing or empty fields, and that can be denoted with a `?` just like in parameter definition.

File: src/bin/TimeLineView.tsx [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+      assignedLane = laneEndTime.length;
+      laneEndTime.push(0);
+    }
+    laneMap.set(span.spanId, assignedLane);
+
+    const realEnd = span.endTime ?? (span.startTime + 3000);
+    laneEndTime[assignedLane] = Math.max(laneEndTime[assignedLane], realEnd);
+  }
+  return laneMap;
+}
```
Comment (aryanjassal) [inline src/bin/TimeLineView.tsx]: I'm not too sure, but usually utility functions are separated out into a `utils.ts` file. If something is specific to this file, and will not be used anywhere else, then you can keep them inside the same file.

File: src/bin/TimeLineView.tsx [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+  // Round earliest down
+  earliest = Math.floor(earliest / TIME_STEP_MS) * TIME_STEP_MS;
+  const maxLane = Math.max(...laneMap.values());
+  const rowCount = 1 + timeToRow(latest, earliest);
+
+  // Initialize a 2D grid: rowCount rows x (maxLane+1) lanes
+  const grid: string[][] = Array.from({ length: rowCount }, () =>
+    Array(maxLane + 1).fill('   ')
+  );
+
```
Comment (aryanjassal) [inline src/bin/TimeLineView.tsx]: Change array type here too.

File: src/bin/TimeLineView.tsx [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+  if (!spans.length) {
+    return <Text>No spans</Text>;
+  }
+
+  const laneMap = assignLanes(spans);
+  let earliest = Math.min(...spans.map(s => s.startTime));
+  let latest = Math.max(...spans.map(s => s.endTime ?? (s.startTime + 3000)));
+
+  // Round earliest down
+  earliest = Math.floor(earliest / TIME_STEP_MS) * TIME_STEP_MS;
```
Comment (aryanjassal) [inline src/bin/TimeLineView.tsx]: You don't change `latest` so that can be `const`. For the flooring operation, you can either use the `Math.floor()` method or the `(earliest / TIME_STEP_MS) * TIME_STEP_MS`. Using both is doing nothing.

You can combine the entire line into this.

`const earliest = Math.floor(Math.min(...spans.map((s) => s.startTime)));`

**Comment by CMCDragonkai** [comment]
Add comments rather than editing spec for that stuff.

File: src/bin/cli.tsx [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+const SPAN_FILE = 'spans.json';
+
+
+const sampleArgIndex = process.argv.indexOf("--sample");
+const sampleMode =
+  sampleArgIndex !== -1 && process.argv.length > sampleArgIndex + 1
+    ? process.argv[sampleArgIndex + 1].trim() === "logical"
+      ? "logical"
+      : "time"
+    : "time";
```
Comment (aryanjassal) [inline src/bin/cli.tsx]: This is fine for testing, but if this is meant to become a part of the codebase permanently, then use a library like `commander.js` to handle arguments.

File: src/bin/cli.tsx [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+const sampleArgIndex = process.argv.indexOf("--sample");
+const sampleMode =
+  sampleArgIndex !== -1 && process.argv.length > sampleArgIndex + 1
+    ? process.argv[sampleArgIndex + 1].trim() === "logical"
+      ? "logical"
+      : "time"
+    : "time"; 
+
+console.log(`Received CLI arguments: ${process.argv.join(" ")}`);
+console.log(`Running in ${sampleMode} mode`);
```
Comment (aryanjassal) [inline src/bin/cli.tsx]: Having `console` statements in your code will raise an error. If this is temporary, you can add an eslint rule to ignore it, but if it it is meant to be permanent, then use the logger methods themselves, or `process.write` to write directly to stdout.

File: src/bin/cli.tsx [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+    }, 1000);
+
+    const handleExit = () => {
+      console.log("Stopping CLI...");
+      clearInterval(id);
+      process.exit(0);
+    };
+
+    process.on("SIGINT", handleExit);
+    process.on("SIGTERM", handleExit);
```
Comment (aryanjassal) [inline src/bin/cli.tsx]: Normally we trap 4 signals, SIGINT, SIGTERM, SIGHUP, and SIGABRT. There is one kill signal we should never trap - SIGKILL.

File: src/lib/span.ts [hunk 19]
```diff
@@ -0,0 +1,38 @@
+export class Span {
+    spanId: string;
+    name: string;
+    startTime: number;
+    endTime: number | null;
+    parentSpanId: string | null;
+    children: Span[];
```
Comment (aryanjassal) [inline src/lib/span.ts]: Review the types here and make sure they follow the standard guidelines.

File: src/lib/span.ts [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+        return {
+            spanId: this.spanId,
+            name: this.name,
+            startTime: this.startTime,
+            endTime: this.endTime,
+            parentSpanId: this.parentSpanId,
+            isCompleted: this.isCompleted(),
+            children: this.children.map(child => child.toJSON())
+        };
+    }
```
Comment (aryanjassal) [inline src/lib/span.ts]: Avoid using the `any` keyword as much as possible. You can look into creating a general `JSONObject` type for the values of the record.

File: src/lib/tracingManager.ts [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+        2
+    );
+}
+
+
+
+
+
+
+
```
Comment (aryanjassal) [inline src/lib/tracingManager.ts]: The linter should have caught these extra spaces. If not, then remove them from the end of the file.

**Review by aryanjassal — CHANGES_REQUESTED** [review]
Most changes are just code style changes. As a summary:
- use `undefined` for null-ish fields
- check if a value is undefined or not by doing a `== null` or `!= null` check (yes, even though we have `undefined` as the type)
- use single quotes
- ensure import formatting
- for array types, always wrap it in `Array<>`
- for potentially undefined fields use `?`, unless you are unable to use `?` for the parameter type, in which case you always use `Type | undefined` (as opposed to `null`)

**Comment by CMCDragonkai** [comment]
@Abby010 can you:

1. Get the CI working
2. Start integrating this into the `js-logger` `src` code under a new class object `Tracer`.
3. Plan out integration into Polykey-CLI for testing.
4. Get @aryanjassal or @tegefaulkes to review to see how it can hook into object lifecycles and node native async contexts.

I think it's important for testing why seed nodes are crashing.

Also why use `-+` and not `\`?

**Comment by CMCDragonkai** [comment]
Also you should be assigning yourself.

**Comment by CMCDragonkai** [comment]
Start adding tests too.

**Comment by CMCDragonkai** [comment]
And benchmarks.

I want a plan as of next cycle.

File: src/lib/span.ts [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+
+    constructor(name: string, parentSpanId: string | null = null) {
+        this.spanId = `span-${Date.now()}-${Math.random().toString(36).substr(2, 5)}`;
+        this.name = name;
+        this.startTime = Date.now();
+        this.endTime = null;
+        this.parentSpanId = parentSpanId;
+        this.children = [];
+    }
+
```
Comment (tegefaulkes) [inline src/lib/span.ts]: The only things that need to be set in the constructor are things that need the be passed through the constructor. Any other property can be defaulted when defined above with `public startTime: number = Date.now()`

The `spanId` should be an `IdSortable` from `@matrixai/id` library.

File: src/lib/span.ts (hunk 19 above)
Comment (tegefaulkes) [inline src/lib/span.ts]: Properties of classes should have an explicit `public` or `protected` specified. These are all public so use `public spanId: string` etc etc.

Also if they're meant to be private then always use the `protected` keyword unless they're meant to be extended in some way.

File: src/lib/span.ts [hunk 23]
```diff
@@ -0,0 +1,38 @@
+export class Span {
```
Comment (tegefaulkes) [inline src/lib/span.ts]: Classes that share the name of the file should use a default export.

```ts
class Span {}

export default Span;

```

All exports are at the bottom of the file.

File: src/lib/span.ts
Comment (tegefaulkes) [inline src/lib/span.ts:1]: This file defines a class, it should be named identically to the class and is case sensitive. So it should be called `Span.ts`.

File: src/lib/span.ts [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+        this.children = [];
+    }
+
+    close(): void {
+        this.endTime = Date.now();
+    }
+
+    isCompleted(): boolean {
+        return this.endTime !== null;
+    }
```
Comment (tegefaulkes) [inline src/lib/span.ts]: All function definitions should have the `public` or `protected` keyword explicitly provided.

File: src/bin/TimeLineView.tsx (hunk 13 above)
Comment (Abby010) [reply src/bin/TimeLineView.tsx]: They are only relevant to the timeline view

**Comment by CMCDragonkai** [comment]
Make sure that the span structure can be streamed into an incremental jsonl format. This is what we will be using to then stream visualise on the browser. @shafiqihtsham can weigh in if you talk to him about it.

But I want you to write out a plan for implementation in Polykey-CLI by today.

**Comment by Abby010** [comment]
# Tracing Integration Plan

## Phase 1: Integrate Tracing into js-logger as a First-Class Component

- Introduce a `Tracer` class that encapsulates all span-related logic.
- Replace the standalone `openSpan`, `closeSpan` functions with methods on this class.
- Support structured span creation: `startSpan`, `endSpan`, and `traced` wrappers with context tracking.
- Provide methods for exporting span data:
  - `getActiveSpans()`
  - `flush()`
  - `getTraceJSON()`

### Add Incremental Span Streaming to JSONL

- Each span, when closed, is written as a JSON object to a `.jsonl` file (e.g., `spans.jsonl`).
- Enable atomic streaming via `fs.appendFileSync` or streaming writers.
- Supports real-time debugging and integration with external tools.

- Update the current TUI visualizer to consume spans from the new `Tracer` instance.

> This phase formalizes span management, embeds JSONL observability, and prepares for broader CLI and system integration.

---

## Phase 2: Integrate Tracer with Polykey CLI and Lifecycle Events

- Initialize and embed the `Tracer` into the Polykey CLI runtime.
- Use `Tracer.traced()` to instrument top-level CLI commands and subcommands.
- Extend tracing into object lifecycle tracking: vaults, daemons, agents, etc.
- Hook into Node.js async context (`AsyncLocalStorage`) to propagate spans implicitly across async calls and promise chains.
- Stream span data from real Polykey CLI runs to the `.jsonl` file.
- Analyze and visualize spans in the TUI to confirm accurate instrumentation.
- Use the streamed data for:
  - Automated debugging
  - Replay analysis
  - Bug discovery

> This phase validates the integration by capturing real CLI behaviors and preparing for long-term observability and tracing.

**Comment by CMCDragonkai** [comment]
Your phase 2 doesn't mention https://github.com/MatrixAI/js-async-init at all. You both need to talk about the implementation plan with @tegefaulkes since I think you're missing some contextual details critical to understanding how tracing can be applied to Polykey's runtime.

**Comment by Abby010** [comment]
# Phase 2 Plan: Polykey CLI Tracing Integration

## Objective

Instrument the Polykey CLI runtime using the Tracer to support structured, memory-safe, and streamable observability of system behavior — with minimal invasiveness and full reversibility.

---

## 1. Embed Tracer into Polykey CLI Runtime

- Inject a `Tracer` instance during CLI bootstrap.
- Pass it into relevant components (e.g., `Agent`, `VaultManager`) via dependency injection or service context.
- Ensure span data is optional and tracing is enabled only via flags like `--enable-tracing`.

---

## 2. Hook into Lifecycle Events via js-async-init

Utilize @asyncInit decorators and structured lifecycle events defined in the [js-async-init](https://github.com/MatrixAI/js-async-init) library.

For async-initialized components like NodeConnectionManager, VaultManager, Daemon, etc., hook into their lifecycle using:

- `EventStart`, `EventStop`
- `EventCreate`, `EventDestroy`

**Instrumentation Pattern:**

```ts
nodeManager.addEventListener(events.NodeConnectionManagerStart.name, () => {
  nodeManager.span = parentSpan.getChild('NodeManagerStart');
});

nodeManager.addEventListener(events.NodeConnectionManagerStop.name, () => {
  nodeManager.span.end();
});
```

## 3. Adopt Span Object Model (No Global Tracking)
Replace global tracer calls with direct Span object usage.

Example:
```
const span = parentSpan.getChild('VaultFetch');
span.event('RPC Request Sent');
span.end();
```

Track spans via objects, not IDs. Avoid propagating span IDs through business logic.

## 4. Stream Output Instead of Writing Directly
Refactor Tracer to emit to a writable stream (WritableStream, PassThrough, etc.).

Consumers of the stream (e.g., file writer, stderr, custom logger) decide how to handle the data.

CLI Support:

```
--trace-output spans.jsonl
--trace-output stderr
--trace-fd=3    # for advanced use
```

## 5. Memory Management: No activeSpans, Use WeakRefs

Remove Map<string, Span> used for global span tracking. 
Span objects should be ephemeral and garbage collected naturally.

Use WeakRef where necessary for parent-child relationships.
Prevent span graphs from being retained in memory permanently.

## 6. Implement Observable Tracing Events
Tracer emits observable events:

span:start
span:end

Useful for real-time debugging, TUI enhancements, or plugin systems.

## 7. Testing Strategy: Use fast-check for Randomized Span Behavior
Replace Math.random() in test harnesses with fast-check property-based fuzzing.

Supports reproducible failures and automated test shrinking for minimal failing cases.

## 8. Visual Debugging with TUI (as-is)
No changes needed to TUI.

Continues to read from the .jsonl file.

## 9.Leveraging js-async-init's Lifecycle Contracts

Use lifecycle annotations from js-async-init to instrument long-lived services and ephemeral agents with start/stop or create/destroy lifecycles. The Tracer will hook into lifecycle phases without modifying the business logic directly.

Each lifecycle boundary maps naturally to a span (e.g., VaultManagerStart, DaemonCreate, NodeConnectionDestroy).
This ensures:

1. Lifecycle spans align with runtime semantics.
2. Tracing remains declarative and reversible.
3. No modification to object constructors or internals.

**Comment by CMCDragonkai** [comment]
Some things to review:

* [https://github.com/tc39/proposal-async-context](https://github.com/tc39/proposal-async-context)
* Review what runtimes support it already (node already does)
* The "span" object should be more implicit, it explore the idea of using WeakMap and symbols for synthetic identifiers.
* Right now committing to an integration path doesn't make sense, try monkey patching at the Polykey runtime itself, and just output logs and visualising that separately.

**Comment by CMCDragonkai** [comment]
Reviewing https://github.com/tc39/proposal-async-context shows that:

1. https://github.com/tc39/proposal-async-context https://github.com/w3c/tpac2023-breakouts/issues/39 https://nodejs.org/api/async_context.html all indicate that this feature isn't available widely in all of JS.
2. If we were to use async context, the logger would be forced to only work in nodejs runtime.
3. This is fine for debugging PK, but not so much crossing into browsers and we want this library to be generic.
4. I realise that we should be able to "approximate" this feature by using our decorators, we have `js-contexts` https://github.com/matrixai/js-contexts that brings together a generic way of creating decorators providing context. It should be possible to use that along with js-async-init to create some sort of more implicit system, that can at first use the node's async context when available, or use something a bit more manual when it is not available (like a combination of weakmap and symbols).

@Abby010 feedback?

**Comment by CMCDragonkai** [comment]
image

**Comment by CMCDragonkai** [comment]
Key thing about this is that our spans don't necessarily end to be useful.

**Comment by Abby010** [comment]
> Reviewing https://github.com/tc39/proposal-async-context shows that:
> 
> 1. https://github.com/tc39/proposal-async-context [`AsyncContext` integration with web APIs w3c/tpac2023-breakouts#39](https://github.com/w3c/tpac2023-breakouts/issues/39) https://nodejs.org/api/async_context.html all indicate that this feature isn't available widely in all of JS.
> 2. If we were to use async context, the logger would be forced to only work in nodejs runtime.
> 3. This is fine for debugging PK, but not so much crossing into browsers and we want this library to be generic.
> 4. I realise that we should be able to "approximate" this feature by using our decorators, we have `js-contexts` https://github.com/matrixai/js-contexts that brings together a generic way of creating decorators providing context. It should be possible to use that along with js-async-init to create some sort of more implicit system, that can at first use the node's async context when available, or use something a bit more manual when it is not available (like a combination of weakmap and symbols).
> 
> @Abby010 feedback?


**Re: async_context Proposal**
I’ve reviewed [tc39/proposal-async-context](https://github.com/tc39/proposal-async-context) and the Node.js [async_hooks](https://nodejs.org/api/async_context.html) API. While these are promising for implicit span propagation, they’re not yet cross-platform.

Agreed for now, relying solely on Node-specific behavior would limit portability. Since js-logger should remain runtime-agnostic, we shouldn’t hard-bind to Node.js-only primitives.

**Re: Span Should Be More Implicit**
Makes sense. Right now, spans are passed explicitly or manually linked via getChild(). I'm exploring using:

1. Symbols to tag async flows.
2. WeakMap to manage span-local state for async resources as suggested
3. Decorators (js-contexts + js-async-init) to set up span context implicitly without forcing manual tracing at every step.

This would allow something like:

```
@Traceable
async function syncVault() {
  // span is implicitly available via decorator-generated context
  currentSpan.event('Vault Sync Triggered');
}
```

**Re: Monkey Patching in Polykey Runtime**
That aligns with the Phase 2 goal: non-invasive, reversible tracing.
Before embedding tracing into the CLI codebase, We will monkey-patch lifecycle events in runtime modules (like VaultManager, NodeConnectionManager) and observe the emitted span logs via JSONL/TUI.

This will let us test the conceptual model before formalizing APIs or integrations.

**Re: js-contexts + js-async-init Hybrid**
I’m looking into combining js-async-init lifecycle events with js-contexts to simulate AsyncLocalStorage-like behavior generically.
Idea is to support both:
Native async context via Node.js hooks.
Fallback shim using js-contexts and WeakRefs where native async context isn't available.

File: tests/asciinemaTest.ts [hunk 25]
```diff
@@ -0,0 +1,148 @@
...
+  hasOrphan: fc.boolean(),
+});
+
+setInterval(async () => {
+  switch (step) {
+    case 0: {
+      current.flags = fc.sample(flagArb, 1)[0];
+      current.parentId = openSpan(`Parent-${parentIndex}`);
+      nestedIds = [];
+      break;
```
Comment (aryanjassal) [inline tests/asciinemaTest.ts:68]: Instead of sampling like this, I would recommend doing something like taking the current Unix time and using that as the seed, and increase the value by one every time you want to sample something new. You should also print the seed value to the console. This will ensure that you are using randomised values every time, and if something goes wrong, you can copy-paste the seed to get the same values to investigate why things went wrong. You should fix an error which happens once in a few runs, as that is an edge case you need to consider instead of rerunning the test and ignoring it.

Otherwise you've done pretty good integrating fast-check into this test.

**Comment by CMCDragonkai** [comment]
The original plan was for @aryanjassal to start using this in Polykey regardless of the merge. But if that's too difficult. Start merging this in now incrementally and release new ones so that PK can start testing it. This week it has to be done. There's no more time to be allocated to this.

**Comment by CMCDragonkai** [comment]
@aryanjassal you now take over this PR to force it to be merged and released.

**Comment by CMCDragonkai** [comment]
@aryanjassal start squashing and merge into a patch version, then use it in a new PR in Polykey-CLI.

File: spans.json
Comment (aryanjassal) [reply spans.json:1]: This is basically done, so I'm resolving this thread.

**Comment by aryanjassal** [comment]
Due to not having sufficient time for the handover from abby, there might be some minor issues which might make it to staging. This might see a lot of iteration as well. All the iteration will be done in patch updates which will be used by Polykey CLI. 

The `js-async-init` package will be monkey-patched first to get kind of a demo of the library, and iteration will be made to this library until we have reached a good state to analyse this in-depth.

**Comment by aryanjassal** [comment]
I've cleaned up the codebase a little bit, so this can now be merged to staging. I will need to go over some things with abby about the future plans of the tracer, but until then I can start testing out monkey-patching.

It doesn't make much sense to make a PR for this as the change will be in node modules, which is local-only and making a PR will not make a difference. I will however make a branch for this and push up a PR if any significant changes are made in there.
